### ISSUE TITLE
Lower training time consistently beats higher training time

### ISSUE THREAD

**Issue by jhmenke** [awslabs/autogluon#840]
Hi, 

first of all, Autogluon is really awesome, i thank you so much for providing such a wonderful tool. 
My use case is relatively simple: i do a tabular prediction with a low number of features (about 30) and -- depending on the use case -- about 10k to 40k samples. This is the script i use:

```
target = "target_col"
time_limit = 3600
eval_metric = "log_loss"
problem_type = "binary" 
predictor = task.fit(train_data=df_train,
                     time_limits=time_limit,
                     eval_metric=eval_metric,
                     presets=['good_quality_faster_inference_only_refit', 'optimize_for_deployment'],
                     problem_type=problem_type,
                     label=target,
                     output_directory=save_dir)

y_pred = predictor.predict_proba(df_test)
loss = log_loss(y_test, y_pred, labels=sorted(list(set(y_test.tolist()))))
```

The weird thing is, that if i set time_limits to one hour, the trained ensemble consistently outperforms much higher time limit (e.g., 12 hours or 24 hours) on the same data. I evaluate the performance with a holdout set (*df_test*). *eval_metric* is also log_loss in the example. I also test with another target (regression), which yields the same results. The underlying samples will be very different, the result the same (less training time = better validation score).

Can anyone make sense of what could be the cause of this weird behavior?

**Comment by Innixma** [comment]
This is very dataset dependent and applies to any ML process that tries to optimize based on training data: It can overfit.

Generally, AutoGluon is very resilient to overfitting, but it is possible to occur, especially if your hold-out data is not representative of your training data. We have a large scale benchmark that we run frequently for autogluon which shows AutoGluon **on average** improves significantly with more time. One thing I do notice is that you are using `good_quality`. If you were using `best_quality` my guess is that it would continue to improve or at the very least not get worse. `good_quality` doesn't gain much after a certain point from more training time as it must simplify the final result to satisfy the latency constraint of the `good_quality` preset.

**Comment by jhmenke** [comment]
Thanks for responding so quickly, that is a good explanation. My holdout set should be representative, but to be honest i have not done an in-depth analysis so it may not be. However, i generally generate new datasets (part of it by shuffling) and the behavior seems consistent (lower time = better results).

I will try the best_quality setting, thanks!
However, this setting will not be realistic for my "production" use, since it generates about 30 GB of models from my 10 MB dataset. I found that the few hundred MB generated by the "good_quality" setting seem a good trade-off.
Is it known why the models get so large? I frequently use catboost, lightgbm and the sklearn-based estimators as well and they never use anything close to that size on disk when pickled.

**Comment by Innixma** [comment]
`best_quality` uses a technique called 'repeated bagging'. It trains many copies of each model on different splits of the data. This leads to improved results but also to diminishing returns relative to the number of copies. In its most extreme, this can train 200 of each model, and the final predictions are the averaged predictions of all 200 models, for each of the model types. This means in the most extreme scenario, `best_quality` can take 200x more space than `good_quality`. Also, it is likely not LGBM/CatBoost that are taking up much space, but instead KNN and RF/XT. You can check for yourself in the output directory where the models are saved to find which models take up the most space.

**Comment by jhmenke** [comment]
Thank again! I think that solves my question(s).
